NorthstarDLL/mods/modsavefiles.cpp

```diff
@@ -0,0 +1,532 @@
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+
+	fs::path dir = savePath / fs::path(mod->m_ModDirectory).filename();
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (!IsPathSafe(fileName, dir))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format(
+				"File name invalid ({})! Make sure it does not contain any non-ASCII character, and results in a path inside your mod's "
+				"save folder.",
+				fileName,
+				mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	// Note - this cannot be done in the async func since the table may get garbage collected.
+	// This means that especially large tables may still clog up the system.
+	std::string content = EncodeJSON<context>(sqvm);
+	if (ContainsInvalidChars(content))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm, fmt::format("File contents may not contain NUL/\\0 characters! Make sure your strings are valid!", mod->Name).c_str());
+		return SQRESULT_ERROR;
+	}
+
+	fs::create_directories(dir);
+	// this actually allows mods to go over the limit, but not by much
+	// the limit is to prevent mods from taking gigabytes of space,
+	// this ain't a cloud service.
+	if (GetSizeOfFolderContentsMinusFile(dir, fileName) + content.length() > MAX_FOLDER_SIZE)
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format(
+				"The mod {} has reached the maximum folder size.\n\nAsk the mod developer to optimize their data usage,"
+				"or increase the maximum folder size using the -maxfoldersize launch parameter.",
+				mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	g_pSaveFileManager->SaveFileAsync<context>(dir / fileName, content);
+
+	return SQRESULT_NULL;
+}
+
+// int NS_InternalLoadFile(string file)
+ADD_SQFUNC("int", NS_InternalLoadFile, "string file", "", ScriptContext::SERVER | ScriptContext::CLIENT | ScriptContext::UI)
+{
+	Mod* mod = g_pSquirrel<context>->getcallingmod(sqvm, 1); // the function that called NSLoadFile :)
+	if (mod == nullptr)
+	{
+		g_pSquirrel<context>->raiseerror(sqvm, "Has to be called from a mod function!");
+		return SQRESULT_ERROR;
+	}
+
+	fs::path dir = savePath / fs::path(mod->m_ModDirectory).filename();
+	std::string fileName = g_pSquirrel<context>->getstring(sqvm, 1);
+	if (!IsPathSafe(fileName, dir))
+	{
+		g_pSquirrel<context>->raiseerror(
+			sqvm,
+			fmt::format(
+				"File name invalid ({})! Make sure it does not contain any non-ASCII character, and results in a path inside your mod's "
+				"save folder.",
+				fileName,
+				mod->Name)
+				.c_str());
+		return SQRESULT_ERROR;
+	}
+
+	g_pSquirrel<context>->pushinteger(sqvm, g_pSaveFileManager->LoadFileAsync<context>(dir / fileName));
+
+	return SQRESULT_NOTNULL;
+}
+
+// bool NSDoesFileExist(string file)
+ADD_SQFUNC("bool", NSDoesFileExist, "string file", "", ScriptContext::SERVER | ScriptContext::CLIENT | ScriptContext::UI)
```

Comment:
Is this necessary? It's less racy to just handle a missing file while attempting to read it instead of having a specific function for it.

Reply:
Modders might want to make use of it in some way, one case i can think of is checking if the user already has saved data for the mod (different from empty saved data) to do some first-time setup or defaults or something.

If it doesn't see use though, it doesn't see use. I personally could go either way with this tbh

Reply:
> checking if the user already has saved data for the mod

Attempt to read it, then if you get a result saying it doesn't exist, do the setup logic. Anything else is just asking for hard-to-debug issues at some point.

Reply:
Idk how I feel about the competing concepts here (prior checking vs handling error cases)

Generally I prefer prior checks to ensure that I won't do weird things, but for squirrel i tend to lean the other way (scripting moment)

I have no preference here overall tbh

Reply:
It's usually better to handle the error cases, unless you don't even plan to open the file to begin with (in which case checking if it exists only is faster and less error-prone). For first-time initialization, error handling is much better in general.

Reply:
I don't see a valid case where a modder will want to check that a file exists without then reading it (except maybe to avoid overriding a file? But if that's the case why weren't they reading it?), so with that being said I'm going to lean on the side of error handling

Reply:
I currently do this but thats mostly because i didn't want to use a try-catch and the failcallback wasn't a thing when i wrote the code that uses that
```
void function cl_titanFrameworkPersistenceInitThread()
{
    if(!NSDoesFileExist("frameworkCFG.json"))
    {
        print("Creating new frameworkCFG.json")
        NSSaveJSONFile("frameworkCFG.json", {version = FRAMEWORK_VERSION});
    }
    if(!NSDoesFileExist("frameworkXP.json"))
    {
        print("Creating new frameworkXP.json")
        NSSaveJSONFile("frameworkXP.json", {version = FRAMEWORK_VERSION});
    }
    NSLoadJSONFile( "frameworkXP.json", loadPersistentXP)
    NSLoadJSONFile( "frameworkCFG.json", loadPersistentData)
}
```

Reply:
> I don't see a valid case where a modder will want to check that a file exists without then reading it (except maybe to avoid overriding a file? But if that's the case why weren't they reading it?), so with that being said I'm going to lean on the side of error handling

1) To confirm overriding a file
2) If a read fails - it may be due to multiple reasons, not just the file not existing.

Reply:
Can this be resolved? I think this function can stay (as it may have usecases)

Reply:
Yeah sure